// @rsbuild/plugin-solid@1.0.6/dist/index.js
import { createRequire } from "node:module";
import { modifyBabelLoaderOptions } from "@rsbuild/plugin-babel";
let src_require = createRequire(import.meta.url), PLUGIN_SOLID_NAME = 'rsbuild:solid';
function pluginSolid(options = {}) {
    return {
        name: PLUGIN_SOLID_NAME,
        setup (api) {
            api.modifyBundlerChain(async (chain, { CHAIN_ID, environment, isProd, target })=>{
                let environmentConfig = environment.config;
                modifyBabelLoaderOptions({
                    chain,
                    CHAIN_ID,
                    modifier: (babelOptions)=>(babelOptions.presets = [
                            [
                                src_require.resolve('babel-preset-solid'),
                                options.solidPresetOptions || {}
                            ]
                        ], babelOptions.parserOpts = {
                            plugins: [
                                'jsx',
                                "typescript"
                            ]
                        }, !isProd && environmentConfig.dev.hmr && 'web' === target && (babelOptions.plugins ??= [], babelOptions.plugins.push([
                            src_require.resolve('solid-refresh/babel')
                        ]), chain.resolve.alias.set('solid-refresh', src_require.resolve('solid-refresh/dist/solid-refresh.mjs'))), babelOptions)
                });
            });
        }
    };
}
export { PLUGIN_SOLID_NAME, pluginSolid };
