// @rsbuild/plugin-babel@1.0.6/dist/index.js
import node_path, { isAbsolute, join, normalize, sep } from "node:path";
import { reduceConfigsWithContext } from "reduce-configs";
import upath from "upath";
import node_fs from "node:fs";
import { createRequire } from "node:module";
import { fileURLToPath } from "node:url";
import deepmerge from "deepmerge";
let BABEL_JS_RULE = 'babel-js', castArray = (arr)=>void 0 === arr ? [] : Array.isArray(arr) ? arr : [
        arr
    ], normalizeToPosixPath = (p)=>upath.normalizeSafe(normalize(p || '')).replace(/^([a-zA-Z]+):/, (_, m)=>`/${m.toLowerCase()}`), formatPath = (originPath)=>isAbsolute(originPath) ? originPath.split(sep).join('/') : originPath, getPluginItemName = (item)=>'string' == typeof item ? formatPath(item) : Array.isArray(item) && 'string' == typeof item[0] ? formatPath(item[0]) : null, modifyPresetOptions = (presetName, options, presets = [])=>{
    presets.forEach((preset, index)=>{
        Array.isArray(preset) ? 'string' == typeof preset[0] && normalizeToPosixPath(preset[0]).includes(presetName) && (preset[1] = {
            ...preset[1] || {},
            ...options
        }) : 'string' == typeof preset && normalizeToPosixPath(preset).includes(presetName) && (presets[index] = [
            preset,
            options
        ]);
    });
}, getBabelUtils = (config)=>{
    let noop = ()=>{};
    return {
        addPlugins: (plugins)=>{
            config.plugins ? config.plugins.push(...plugins) : config.plugins = plugins;
        },
        addPresets: (presets)=>{
            config.presets ? config.presets.push(...presets) : config.presets = presets;
        },
        removePlugins: (plugins)=>{
            ((plugins, config)=>{
                if (!config.plugins) return;
                let removeList = castArray(plugins);
                config.plugins = config.plugins.filter((item)=>{
                    let name = getPluginItemName(item);
                    return !name || !removeList.find((removeItem)=>name.includes(removeItem));
                });
            })(plugins, config);
        },
        removePresets: (presets)=>{
            ((presets, config)=>{
                if (!config.presets) return;
                let removeList = castArray(presets);
                config.presets = config.presets.filter((item)=>{
                    let name = getPluginItemName(item);
                    return !name || !removeList.find((removeItem)=>name.includes(removeItem));
                });
            })(presets, config);
        },
        addIncludes: noop,
        addExcludes: noop,
        modifyPresetEnvOptions: (options)=>{
            modifyPresetOptions('@babel/preset-env', options, config.presets || []);
        },
        modifyPresetReactOptions: (options)=>{
            modifyPresetOptions('@babel/preset-react', options, config.presets || []);
        }
    };
}, modifyBabelLoaderOptions = ({ chain, CHAIN_ID, modifier })=>{
    for (let ruleId of [
        CHAIN_ID.RULE.JS,
        CHAIN_ID.RULE.JS_DATA_URI,
        BABEL_JS_RULE
    ])if (chain.module.rules.has(ruleId)) {
        let rule = chain.module.rule(ruleId);
        rule.uses.has(CHAIN_ID.USE.BABEL) && rule.use(CHAIN_ID.USE.BABEL).tap(modifier);
    }
}, plugin_dirname = node_path.dirname(fileURLToPath(import.meta.url)), plugin_require = createRequire(import.meta.url), PLUGIN_BABEL_NAME = 'rsbuild:babel', SCRIPT_REGEX = /\.(?:js|jsx|mjs|cjs|ts|tsx|mts|cts)$/, DEFAULT_BABEL_PRESET_TYPESCRIPT_OPTIONS = {
    allowNamespaces: !0,
    allExtensions: !0,
    allowDeclareFields: !0,
    optimizeConstEnums: !0,
    isTSX: !0
};
async function getCacheIdentifier(options) {
    let identifier = `${process.env.NODE_ENV}${JSON.stringify(options)}`, { version: coreVersion } = await import("@babel/core"), loaderVersion = JSON.parse(await node_fs.promises.readFile(join(plugin_dirname, '../compiled/babel-loader/package.json'), 'utf-8')).version ?? '';
    return identifier + `@babel/core@${coreVersion}babel-loader@${loaderVersion}`;
}
let getDefaultBabelOptions = (config, context)=>{
    let isLegacyDecorators = 'legacy' === config.source.decorators.version, options = {
        babelrc: !1,
        configFile: !1,
        compact: 'production' === config.mode,
        plugins: [
            [
                plugin_require.resolve('@babel/plugin-proposal-decorators'),
                config.source.decorators
            ],
            ...isLegacyDecorators ? [
                plugin_require.resolve('@babel/plugin-transform-class-properties')
            ] : []
        ],
        presets: [
            [
                plugin_require.resolve("@babel/preset-typescript"),
                DEFAULT_BABEL_PRESET_TYPESCRIPT_OPTIONS
            ]
        ]
    }, { buildCache = !0 } = config.performance;
    if (buildCache && 'rspack' === context.bundlerType) {
        var cacheDirectory;
        let cacheDirectory1 = (cacheDirectory = 'boolean' == typeof buildCache ? void 0 : buildCache.cacheDirectory) ? isAbsolute(cacheDirectory) ? cacheDirectory : join(context.rootPath, cacheDirectory) : join(context.cachePath);
        options.cacheCompression = !1, options.cacheDirectory = join(cacheDirectory1, 'babel-loader');
    }
    return options;
}, pluginBabel = (options = {})=>({
        name: PLUGIN_BABEL_NAME,
        setup (api) {
            let getBabelOptions = async (environment)=>{
                let { config } = environment, mergedOptions = ((defaultOptions, userBabelConfig, extraBabelUtils)=>{
                    if (userBabelConfig) {
                        let babelUtils = {
                            ...getBabelUtils(defaultOptions),
                            ...void 0
                        };
                        return reduceConfigsWithContext({
                            initial: defaultOptions,
                            config: userBabelConfig,
                            ctx: babelUtils
                        });
                    }
                    return defaultOptions;
                })(deepmerge({}, getDefaultBabelOptions(config, api.context)), options.babelLoaderOptions);
                return mergedOptions.cacheDirectory && !mergedOptions.cacheIdentifier && (mergedOptions.cacheIdentifier = await getCacheIdentifier(mergedOptions)), mergedOptions;
            };
            api.modifyBundlerChain({
                order: 'pre',
                handler: async (chain, { CHAIN_ID, environment })=>{
                    let babelOptions = await getBabelOptions(environment), babelLoader = node_path.resolve(plugin_dirname, '../compiled/babel-loader/index.js'), { include, exclude } = options;
                    if (include || exclude) {
                        let rule = chain.module.rule(BABEL_JS_RULE).after(CHAIN_ID.RULE.JS);
                        if (include) for (let condition of castArray(include))rule.include.add(condition);
                        if (exclude) for (let condition of castArray(exclude))rule.exclude.add(condition);
                        rule.test(SCRIPT_REGEX).use(CHAIN_ID.USE.BABEL).loader(babelLoader).options(babelOptions);
                    } else chain.module.rule(CHAIN_ID.RULE.JS).test(SCRIPT_REGEX).use(CHAIN_ID.USE.BABEL).after(CHAIN_ID.USE.SWC).loader(babelLoader).options(babelOptions);
                }
            });
        }
    });
export { PLUGIN_BABEL_NAME, getBabelUtils, getDefaultBabelOptions, modifyBabelLoaderOptions, pluginBabel };
